src/StepperMotorReadback.cc

```diff
@@ -7,118 +7,175 @@
+    deviceError.message = "";
+    deviceError.setCurrentVersionNumber({});
+    deviceError.writeAll();
 
     while(true) {
       //First calculate the initial values
-
-      // FIXME This is only to evaluate the timer
-      if(!execTimer.isInitialized()) {
-        execTimer.initializeMeasurement();
+      try {
+        // FIXME This is only to evaluate the timer
+        if(!execTimer.isInitialized()) {
+          execTimer.initializeMeasurement();
+        }
+        else {
+          execTimer.measureIterativeMean();
+        }
+        auto ct = std::chrono::duration_cast<std::chrono::microseconds>(execTimer.getMeasurementResult());
+        actualCycleTime = static_cast<float>(ct.count()) / 1000.f;
+
+        receiveTimer.initializeMeasurement();
+
+        readback();
+        if(_motor->get()->hasHWReferenceSwitches()) {
+          readEndSwitchData();
+        }
+
+        receiveTimer.measureOnce();
+        auto rt = std::chrono::duration_cast<std::chrono::microseconds>(receiveTimer.getMeasurementResult());
+        actualReceiveTime = static_cast<float>(rt.count()) / 1000.f;
+
+        //We now have all data. Write them
+        writeAll();
+
+        // We have aparently communicated successfully. Reset the error counter
+        spiErrorCounter = 0;
       }
-      else {
-        execTimer.measureIterativeMean();
+      catch(mtca4u::MotorDriverException& e) {
+        if((e.getID() == mtca4u::MotorDriverException::SPI_ERROR ||
+               e.getID() == mtca4u::MotorDriverException::SPI_TIMEOUT)) {
+          if(spiErrorCounter > SPI_RETRY_COUNT) {
+            std::cerr << "SPI failed..." << std::endl;
+            incrementDataFaultCounter();
+            _motor->close();
+            deviceError.status = static_cast<int32_t>(StatusOutput::Status::FAULT);
+            deviceError.message = e.what();
+            deviceError.setCurrentVersionNumber({});
+            deviceError.writeAll();
+          }
+          else {
+            spiErrorCounter++;
+          }
+        }
+        else {
+          throw std::current_exception();
+        }
       }
-      auto ct = std::chrono::duration_cast<std::chrono::microseconds>(execTimer.getMeasurementResult());
-      actualCycleTime = static_cast<float>(ct.count()) / 1000.f;
-
-      receiveTimer.initializeMeasurement();
 
-      readback();
-      if(_motor->hasHWReferenceSwitches()) {
-        readEndSwitchData();
+      if(not _motor->isOpen()) {
+        std::cerr << "Starting device recovery for " << _motor->toString() << std::endl;
+        do {
+          try {
+            _motor->renew();
+          }
+          catch(mtca4u::MotorDriverException& e) {
+            if(std::string(deviceError.message) != e.what()) {
+              deviceError.message = e.what();
+              deviceError.setCurrentVersionNumber({});
+              deviceError.message.write();
+            }
+          }
+          boost::this_thread::sleep_for(boost::chrono::milliseconds(10));
```

Comment:
Do you really want to try re-opening the motor with 100 Hz? Why not have one shot with the main cyclic trigger?

Reply:
Without having initial values first and the trigger last I would write it like this
```
while(true){
  trigger.read();
  if (!isOpen){
    try{
     motor->open();
    }catch(MotorException &){
      continue;
    }
  }
  try{
     myBusinessLogic();
    }catch(MotorException &){
      motor->close();
    }
  }
}
```
Let me see if I can get this with "initial values first"....
```
while(true){
  if (isOpen){
    try{
     someStuffThatMightThrow();
     writeAll();
    }catch(MotorException&){
      motor->close();
    }
  }
  trigger.read();
  if (!isOpen()){
    try{
      motor->open();
    }catch(MotorException){ }
  }

}
```

Reply:
Another problem I see is that the writeAll() only happens at the end of the first try-block. If there is a problem, the module never writes the initial values.

Reply:
> Do you really want to try re-opening the motor with 100 Hz? Why not have one shot with the main cyclic trigger?

I somehow misread the 500000 usec in DeviceModule as 10000

Reply:
I had tried to put it that way but could not get it to work, so I went and modelled it after the recovery in DeviceModule

Reply:
Changed. I have moved the writeAll() after the try block. If there is an error during the initial value phase, we flush out initial values that are marked invalid that way.